### PR TITLE
BUG: fix inserting tz-aware datetime to Series, closes #12862

### LINKED ISSUE #12862: Series.__setitem__ with datetime64[ns, tz] dtype raises ValueError when setting a tz-aware timestamps
I'm not able to construct a Series element by element if the elements are timezone localized datetimes. Any workarounds?
#### Code Sample, a copy-pastable example if possible

``` python
# This works
x = pd.Series()
x['foo'] = pd.to_datetime(42)
x['bar'] = pd.to_datetime(666)

# This doesn't
x = pd.Series()
x['foo'] = pd.to_datetime(42).tz_localize('UTC')
x['bar'] = pd.to_datetime(666).tz_localize('UTC')
```

```
ValueError                                Traceback (most recent call last)
<ipython-input-24-cc1ab78086e5> in <module>()
      1 x = pd.Series()
      2 x['foo'] = pd.to_datetime(42).tz_localize('UTC')
----> 3 x['bar'] = pd.to_datetime(666).tz_localize('UTC')

/usr/lib/python3.5/site-packages/pandas/core/series.py in __setitem__(self, key, value)
    726         # do the setitem
    727         cacher_needs_updating = self._check_is_chained_assignment_possible()
--> 728         setitem(key, value)
    729         if cacher_needs_updating:
    730             self._maybe_update_cacher()

/usr/lib/python3.5/site-packages/pandas/core/series.py in setitem(key, value)
    722                     pass
    723 
--> 724             self._set_with(key, value)
    725 
    726         # do the setitem

/usr/lib/python3.5/site-packages/pandas/core/series.py in _set_with(self, key, value)
    770                 self._set_values(key.astype(np.bool_), value)
    771             else:
--> 772                 self._set_labels(key, value)
    773 
    774     def _set_labels(self, key, value):

/usr/lib/python3.5/site-packages/pandas/core/series.py in _set_labels(self, key, value)
    780         mask = indexer == -1
    781         if mask.any():
--> 782             raise ValueError('%s not contained in the index' % str(key[mask]))
    783         self._set_values(indexer, value)
    784 

ValueError: ['b' 'a' 'r'] not contained in the index
```
#### Expected Output

Shouldn't raise errors.
#### output of `pd.show_versions()`

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.6-gnu-1
machine: x86_64
processor: 
byteorder: little
LC_ALL: None
LANG: en_DK.UTF-8

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 18.2
Cython: 0.24
numpy: 1.11.0
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: 1.4
patsy: None
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: None
tables: 3.2.2
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: None
xlrd: 0.9.4
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.4.1
html5lib: None
httplib2: 0.9.2
apiclient: 1.5.0
sqlalchemy: 1.0.12
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.8
boto: None
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #12862
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/series/indexing/test_indexing.py [hunk 1]
```diff
@@ -523,7 +523,22 @@ def test_setitem_with_tz_dst():
     tm.assert_series_equal(s, exp)
 
 
-def test_categorial_assigning_ops():
+def test_setitem_new_key_tz():
```
Comment (jreback) [inline pandas/tests/series/indexing/test_indexing.py]: there is a file for timezones and timeseries tests under indexing

prefer one of those

File: pandas/core/series.py [hunk 2]
```diff
@@ -1230,6 +1230,13 @@ def setitem(key, value):
                 if _is_unorderable_exception(e):
                     raise IndexError(key)
 
+                if is_scalar(key) and not is_integer(key) and key not in self.index:
```
Comment (jreback) [inline pandas/core/series.py]: i would trace further

you are adding a new case whereby it’s likely an existing path is just incorrect somewhere

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/indexing/test_indexing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: sure.  for next time, when does a test go in tests.indexing vs tests.series.indexing vs tests.frame.test_indexing?

Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: both tests.series.indexing.test_datetime and tests.indexing.test_datetime are about objects with DatetimeIndex, not datetime values

File: pandas/core/series.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/series.py]: the path here is above when we call `self._set_with_engine(key, value)`, `_set_with_engine` calls `self.index._engine.set_value(self._values, key, value)`.  The engine expects `self._values` to be an ndarray, but here it is a `DatetimeArray`.  Eventually it _might_ be worth addressing this at the low-level engine level, but definitely out of scope for now.

Comment (mroeschke) [reply pandas/core/series.py]: FWIW, that's where I got stuck trying to fix this same bug. I tried converting it to an ndarray before passing it into the engine but a lot of other routines failed.

File: pandas/tests/series/indexing/test_indexing.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/series/indexing/test_indexing.py]: well test_indexing is really big, just like to split out when dealing with specific types (may not be perfect but that's the idea); this should clearly be there.

File: pandas/tests/series/indexing/test_indexing.py [hunk 3]
```diff
@@ -523,7 +523,22 @@ def test_setitem_with_tz_dst():
...
+    # GH#12862 should not raise on assigning the second value
+    vals = [
+        pd.to_datetime(42).tz_localize("UTC"),
+        pd.to_datetime(666).tz_localize("UTC"),
+    ]
+
```
Comment (jreback) [inline pandas/tests/series/indexing/test_indexing.py]: can you *also* test using .loc which give the same result

Comment (jreback) [reply pandas/tests/series/indexing/test_indexing.py]: we may actually have a test for this already (so good to combine)

File: pandas/tests/series/indexing/test_indexing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: i think figuring this out merits a dedicated pass(es) similar to tests.arithmetic and tests.reductions

Comment (jreback) [reply pandas/tests/series/indexing/test_indexing.py]: maybe, but better to put this where it should go now

Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: this file is only 842 lines long and contains a bunch of tests for setitem.  I still don't see what the preferred alternative is

Comment (jreback) [reply pandas/tests/series/indexing/test_indexing.py]: see my comment below, I am sure we test this for .loc, so this belongs with it.

Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: OK.  Do you have a suggestion as to where that might be?  I'm not eager to manually sort through 50k+ tests.  (and there aren't _any_ `foo.loc` setitem calls in either tests.series.test_timezones or tests.series.test_timeseries)

File: pandas/core/series.py (hunk 2 above)
Comment (jreback) [reply pandas/core/series.py]: move this to _set_with and its better (after the cases are narrowed down)

File: pandas/tests/series/indexing/test_indexing.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/series/indexing/test_indexing.py]: ``test_loc_coerceion`` in pandas/tests/indexing/test_loc.py

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/series.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/series.py]: sure

File: pandas/tests/series/indexing/test_indexing.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: everything in that file is testing loc.  you want to mix non-loc tests in with that?

Comment (jbrockmendel) [reply pandas/tests/series/indexing/test_indexing.py]: moved

**Comment by jbrockmendel** [comment]
comments addressed i think

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ok thanks; this looks better.

**Pull request merged**